### DISCUSSION TITLE
SSG build fails with SyntaxError when processing library with css import

### DISCUSSION THREAD

**Discussion by warmshadow** [vikejs/vike#617] [Discussion]
Hey, started playing around with the plugin. I got a really minimal setup right now, one `index.page.jsx` and `ssr({ prerender: true })` set inside vite config. My project uses library that imports css styles from react toastify. Build without plugin works as expected, but trying to build pre-rendered pages with plugin results in this error:

```
/myProject/node_modules/myLibrary/node_modules/react-toastify/dist/ReactToastify.css:1
:root {
^

SyntaxError: Unexpected token ':'
    at Object.compileFunction (node:vm:360:18)
    .....
```

Any idea what am I missing here?

**Comment by brillout** [comment]
What's the full stack trace?

**Reply by warmshadow** [reply]
```
/myProject/node_modules/myLibrary/node_modules/react-toastify/dist/ReactToastify.css:1
:root {
^

SyntaxError: Unexpected token ':'
    at Object.compileFunction (node:vm:360:18)
    at wrapSafe (node:internal/modules/cjs/loader:1088:15)
    at Module._compile (node:internal/modules/cjs/loader:1123:27)
    at Module._extensions..js (node:internal/modules/cjs/loader:1213:10)
    at Module.load (node:internal/modules/cjs/loader:1037:32)
    at Module._load (node:internal/modules/cjs/loader:878:12)
    at Module.require (node:internal/modules/cjs/loader:1061:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at /myProject/node_modules/myLibrary/dist/index.js:1:23309
    at Object.<anonymous> (/myProject/node_modules/myLibrary/dist/index.js:1:52745)
```

**Reply by brillout** [reply]
Seems like the CSS is being loaded by Node.js. Why is that so?

**Reply by warmshadow** [reply]
If it helps - even after this error I still can see styles from that css file get added to generated `/dist/client/assets/index.page.[hash].css`, no generated html page though as this crash happens on `vite-plugin-ssr v0.4.71 pre-rendering HTML...` build step

**Reply by warmshadow** [reply]
> Seems like the CSS is being loaded by Node.js. Why is that so?

That library in question has generic global css import for those styles `import 'react-toastify/dist/ReactToastify.css';`, which results in build file `/myLibrary/dist/index.js` having `require("react-toastify/dist/ReactToastify.css");` statement.
Now I'm using the same library in my other (Gatsby) project and there's no problem pre-generating pages there, every page gets these styles injected in html file during build by default, probably because of some transformation when loading that css file.

Is this the problem on my side? What could be the right approach to make it work with VPS?

**Reply by brillout** [reply]
In what file of `react-toastify`'s source code does the import statement `import 'react-toastify/dist/ReactToastify.css';` live?

**Reply by brillout** [reply]
Actually, the best would be a minimal reproduction. I'll have a look at it (there are several possible solution but I'd like to provide the most sensible one.)

**Reply by warmshadow** [reply]
No, that import statement is in my own custom library (`myLibrary`) that is a dependency to my vite project

**Reply by brillout** [reply]
If you add `react-toastify` to a VPS boilerplate/example, do you get the same error?

**Reply by warmshadow** [reply]
Ok I solved part of it, but found couple new issues.
As I mentioned it's not a problem with react-toastify itself, it just only happens that it was that specific import in my situation.

First of all this was an oversight on my end as my custom npm package was built only in CommonJS and had no ESM build files. As it worked in other projects and scenarios I didn't think about it right away.
 
I did a fresh test with VPS boilerplate project and to achieve successful pre-rendering of pages either one of two solutions work: 
-  Building ESM version of my library
OR
-  Workaround by adding `ssr: { noExternal: ["myLibrary"]}` to vite.config.ts

BUT. It only works work for pre-rendering pages. 
If i want to run it as SSR it doesn't work, here are couple scenarios:

1. If my package is only CommonJS and I add `ssr: { noExternal: ["myLibrary"]}` then when I run server it works until I visit a page which imports from `myLibrary` - I get `ReferenceError: require is not defined  at /node_modules/myLibrary/dist/index.js:1:740`.
2. Then after building my package as ESM code, when I run the server again and visit the same page, I get error from library's deeper 'react-is' dependency - `ReferenceError: module is not defined at .../react-is/index.js:6:3`. And here it doesn't matter if I add myLibrary to `noExternal `or not.

I'm not sure if I should continue here as it's probably a bit out of scope for this discussion, as it's a different problem. We can move to other discussion if needed. Though my question is - why do you think pre-rendering works fine with this setup and SSR server doesn't handle it? What should be done to fix it, as this time the package related to a problem is out of my reach?

**Reply by brillout** [reply]
There isn't a silver bullet here and in my experience the best is to actually dig and understand what's going on.

ESM/CJS compat is a mess and I'm afraid there isn't much VPS can do to help (although I've a couple of tricks in mind to make VPS workaround/alleviate the compat messiness a bit.)
